### DISCUSSION TITLE
confused by rule

### DISCUSSION THREAD

**Discussion by catdogmaus** [arendst/Tasmota#22107] [Q&A] [answered]
I found from cookbook simple rule that basically should work exactly for me. I think it was for solar heater. I have different heater but that is not important.
Another noteworthy difference is that it uses Sonoff TH10 and I have ESP32D so I have "relay_i 1" on GPIO26. Still as far as I understand it should mean relay1 is power1 by rules?! So this rule should work.
First thing I see when adding below rule is 
``
20:25:46.728` CFG: Text overflow by 170 char(s)
20:25:46.733 MQT: stat/heaterFEC0D0/RESULT = `{"Mem3":""}
``
Probably pointless to say that rule does not work for me. Is the   `power1 0;` wrong in my case?
```
mem3 30

rule1
  on DS18B20-1#temperature do
    event t1=%value%
  endon
  on DS18B20-2#temperature do
    event t2=%value%
  endon
  on event#t2>%mem3% do 
    var1 1;
  endon
  on event#t2<=%mem3% do 
    var1 0;
  endon
  on event#t1 do 
    backlog
    var2 %value%;
    add2 1;
  endon
  on event#t1 do 
    backlog
    var3 %value%;
    add3 2;
  endon
  on event#t2>%var3% do
    power1 %var1%;
  endon
  on event#t2<%var2% do
    power1 0;
  endon

rule1 1
```

**Comment by sfromis** [comment, marked as answer]
I'd guess that you are copy/pasting the above in a single operation, making the whole to be seen as argument to the `Mem` command, overflowing the text pool. This cannot work, each of the 3 commands has to be one copy/pate.

A pin configured as `Relay_i` is just the same as `Relay`, except that the high/low level is inverted on the output, that depends on how the hardware is built. And the first relay will indeed be `Power1`.

**Reply by catdogmaus** [reply]
Indeed, 
This rule does not have error handling for temp values. When I will change the rule, can I just overwrite it or should I erase it first? Same goes for `Mem` value.
Also, do u know is it possible to overrule relay1 if I want to turn pump on manually from HA.
I figured that maybe I need rule2 that turns off rule1 and turns on relay1 for that but I have no idea how relay1 on from HA should trigger that.

**Reply by sfromis** [reply]
In both cases, Just overwriting it will replace it in full, no separate erase needed.

And you can issue `Power1` commands anytime, including from HA, if you want. Beware if this would "confuse" the rules, but you do not need to deactivate them to run the `Power1` command.

If you want to suspend the rule set from operating, the `Rule` command does have an option to enable/disable the rule set, but you could also have a variable, and `if/endif` around parts you want to be able to skip.

**Reply by catdogmaus** [reply]
that is not what I meant exactly. Obviously at moment rule1 will overrule the switch function in HA and also toggle button in esp. It is possible to change that with another rule. Do you know what is the command or trigger that comes from HA to Tasmota when I try to activate the switch in HA? Whatever other rule I am able to create needs to triggered with that. I can not just manipulate with power1 bcs rule1 already does that. I need some trigger.

**Reply by sfromis** [reply]
If you ask about what to do in HA, I've avoided getting intimate with that tool, but the `Power` command is what I'd expect trying to control a `Relay` in Tasmota. The `Event` command can work well for a custom trigger you can create on Tasmota.

**Reply by catdogmaus** [reply]
Can you maybe tell me is this valid? I did add dummy relay as power2. Only thing is this seems not to work 

```
rule2

 on power2 1 do rule1 0 endon
 on power2 0 do rule1 1 endon
```

**Reply by sfromis** [reply]
Not valid, use `Power2#state==0` etc.

**Reply by catdogmaus** [reply]
Thanks, I got now working rules and also functional switch in HA.

I still have hopefully some final questions to get some tasmota rules stuff clearer for me.
Now I have two switches in HA, because tasmota creates switch automatically when relay is added in gpios. When actually with switch - no switch is created. I would like to change unnecessary relay to switch removing it so from manual options but when `relay1` is `power1` what is `switch1` in rules?
Also I have never seen anywhere what does actually mean `relay_b`, `relay_bi` and also `switch_n` and `switch_d`?

**Reply by sfromis** [reply]
Beware of messing up terms. Seen from the Tasmota, a "Switch" is an input pin, has nothing directly to do with a "Relay" which is an output pin, you cannot change pin configs either way. But HA uses the term "switch" for a switchable power *output*, which in Tasmota is a Relay". Unless you configure manually, you can only get a "switch" in HA by having a "Relay" (or other output on/off channels) in Tasmota. In the context of Tasmota, it does not make sense to expect that a Switch input would be reflected as if it was output.

In rules, Switch1 is Switch1, like when Button1 is Button1. And while Relay1 is one of the ways to get a Power1, it is not the only way, as it can also be for a light or display. And you can also switch a power output channel using commands like `Power`, irrespective of whether it is a Relay, PWM, light or display.

The gpio suffixes you mention are `b` for bistable, `i` for inverted, `n` for no pull-up, and `d` for pulldown.
